### PR TITLE
Fix ldap.MOD_ADD, ldap.MOD_DELETE and ldap.MOD_REPLACE operations on multi-value attribute descriptions

### PR DESCRIPTION
**Pull request by sunweaver**

### CODE REVIEW

**Comment by timabbott** [comment]
Can you include tests?

-Tim Abbott (mobile)

**Comment by sunweaver** [comment]
@timabbott unit tests have now been added. I also had to pull in the patch from #20 into this PR to make tests work.

File: fakeldap.py [hunk 1]
```diff
@@ -302,10 +302,18 @@ def _modify_s(self, dn, mod_attrs):
         for item in mod_attrs:
             op, key, value = item
             if op is 0:
-                # FIXME: Can't handle multiple entries with the same name
-                # its broken right now
-                # do a MOD_ADD, assume it to be a list of values
```
Comment (mateuszmandera) [inline fakeldap.py:307]: Can we leave the piece of the comment saying this is MOD_ADD?

File: fakeldap.py [hunk 2]
```diff
@@ -316,13 +316,28 @@ def _modify_s(self, dn, mod_attrs):
...
                 else:
+                    # delete values individually (and see what values remain)
+                    for val in value:
+                        for i in range(len(row)-1, -1, -1):
+                            if val == row[i]:
+                                # FIXME: throw an ldap exception here (not such value or
+                                # simiar, not sure what happens in real life here)
+                                del row[i]
+
+                # if no values remain, drop the entire LDAP attribute description from the LDAP entry
```
Comment (mateuszmandera) [inline fakeldap.py:358]:
```suggestion
# if no values were specified, drop the entire LDAP attribute description from the LDAP entry
```
would be more correct wording I think

**Review by mateuszmandera — APPROVED** [review]
These all look like valuable improvements and the implementation looks correct for `fakeldap` (we're not intending to be covering all the edge cases and accurately mimicking real LDAP behavior for them - and this expands the range of well-handled scenarios), so I think it should be good to merge @timabbott

**Comment by sunweaver** [comment]
On  Sa 17 Jul 2021 16:36:03 CEST, Mateusz Mandera wrote:

> @mateuszmandera approved this pull request.
>
> These all look like valuable improvements and the implementation  
> looks correct for `fakeldap` (we're not intending to be covering all  
> the edge cases and accurately mimicking real LDAP behavior for them  
> - and this expands the range of well-handled scenarios), so I think  
> it should be good to merge @timabbott

I am ok with all suggested improvements. However, I can't do the PR  
update, as I am going on holiday tomorrow early morning.

Please add the suggested comment changes as a commit on top of adjust  
my PR branch (you should be able to force-push to it).

Thanks+Greets,
Mike
-- 

DAS-NETZWERKTEAM
c\o Technik- und Ökologiezentrum Eckernförde
Mike Gabriel, Marienthaler Str. 17, 24340 Eckernförde
mobile: +49 (1520) 1976 148
landline: +49 (4351) 850 8940

GnuPG Fingerprint: 9BFB AEE8 6C0A A5FF BF22  0782 9AF4 6B30 2577 1B31
mail: ***@***.***, http://das-netzwerkteam.de

**Comment by timabbott** [comment]
@mateuszmandera feel free to make that tweak and merge (I just made sure you have access; should be easy with our `reset-to-pull-request` / `push-to-pull-request` tools).

**Comment by mateuszmandera** [comment]
Merged, thanks a lot @sunweaver !

**Pull request closed without merging**